### PR TITLE
Add apply_ to dataclasses

### PR DESCRIPTION
**Pull request by fzimmermann89**
moved from #489 + additional test

### CODE REVIEW

File: src/mrpro/data/MoveDataMixin.py [hunk 1]
```diff
@@ -219,26 +223,64 @@ def _mixin_to(obj: MoveDataMixin) -> MoveDataMixin:
...
+    ) -> Self:
+        """Apply a function to all children in-place.
+
+        Parameters
+        ----------
+        function
+            The function to apply to all fields. None is interpreted as a no-op.
+
+        memo
+            A dictionary to keep track of  objects that the function has already been applied to,
```
Comment (ckolbPTB) [inline src/mrpro/data/MoveDataMixin.py]:
```suggestion
            A dictionary to keep track of objects that the function has already been applied to,
```

File: src/mrpro/data/MoveDataMixin.py [hunk 2]
```diff
@@ -219,26 +223,64 @@ def _mixin_to(obj: MoveDataMixin) -> MoveDataMixin:
...
+        memo: dict[int, Any] | None = None,
+        recurse: bool = True,
+    ) -> Self:
+        """Apply a function to all children in-place.
+
+        Parameters
+        ----------
+        function
+            The function to apply to all fields. None is interpreted as a no-op.
+
```
Comment (ckolbPTB) [inline src/mrpro/data/MoveDataMixin.py]:
```suggestion
```

File: src/mrpro/data/MoveDataMixin.py [hunk 3]
```diff
@@ -219,26 +223,64 @@ def _mixin_to(obj: MoveDataMixin) -> MoveDataMixin:
...
+        memo
+            A dictionary to keep track of  objects that the function has already been applied to,
+            to avoid multiple applications. This is useful if the object has a circular reference.
+
```
Comment (ckolbPTB) [inline src/mrpro/data/MoveDataMixin.py]:
```suggestion
```

File: src/mrpro/data/SpatialDimension.py [hunk 4]
```diff
@@ -109,6 +108,16 @@ def from_array_zyx(
 
         return SpatialDimension(z, y, x)
 
+    def apply_(self, function: Callable[[T], T] | None = None, **_) -> Self:
```
Comment (ckolbPTB) [inline src/mrpro/data/SpatialDimension.py:111]: why do we need `**_`? For later overloads?

**Review by ckolbPTB — APPROVED** [review]

File: src/mrpro/data/SpatialDimension.py (hunk 4 above)
Comment (fzimmermann89) [reply src/mrpro/data/SpatialDimension.py:111]: it overloads the one from MoveDataMixin, which has additional keyword only parameters (recurse and memo), which Ithink don#t make sense here. But we could also have them and ignore them..
The only reason for apply_ here is to define it with a function Callable[[T], T], as the function does not have to deal with any other children-types.

**Pull request merged**